File: worker/src/store/stats.ts
import { DueStatsRow, SendStats, nowISO } from './types';

type SendStatsRow = {
  send_id: string;
  sent: number;
  delivered: number;
  opened: number;
  clicked: number;
  failed: number;
  complained: number;
  unsubscribed: number;
  first_fetched_at: string | null;
  last_fetched_at: string | null;
  next_fetch_at: string | null;
  is_final: number;
  fetch_error: string | null;
  created_at: string;
  updated_at: string;
};

function rowToStats(r: SendStatsRow): SendStats {
  return {
    send_id: r.send_id,
    sent: r.sent ?? 0,
    delivered: r.delivered ?? 0,
    opened: r.opened ?? 0,
    clicked: r.clicked ?? 0,
    failed: r.failed ?? 0,
    complained: r.complained ?? 0,
    unsubscribed: r.unsubscribed ?? 0,
    first_fetched_at: r.first_fetched_at,
    last_fetched_at: r.last_fetched_at,
    next_fetch_at: r.next_fetch_at,
    is_final: r.is_final === 1,
    fetch_error: r.fetch_error,
    created_at: r.created_at,
    updated_at: r.updated_at,
  };
}

export async function getSendStats(db: D1Database, sendID: string): Promise<SendStats | null> {
  const row = await db
    .prepare(
      `SELECT send_id, sent, delivered, opened, clicked, failed, complained, unsubscribed,
              first_fetched_at, last_fetched_at, next_fetch_at, is_final, fetch_error,
              created_at, updated_at
         FROM send_stats WHERE send_id = ?`,
    )
    .bind(sendID)
    .first<SendStatsRow>();
  return row ? rowToStats(row) : null;
}

export function initSendStatsStmt(db: D1Database, sendID: string): D1PreparedStatement {
  return db.prepare(`INSERT OR IGNORE INTO send_stats (send_id) VALUES (?)`).bind(sendID);
}

export function markSendCompletedForStatsStmt(
  db: D1Database,
  sendID: string,
): D1PreparedStatement {
  // next_fetch_at must be written in the SAME format everywhere (JS ISO-8601,
  // e.g. "2026-06-05T12:09:15.905Z") so listDueSendStats can compare it as a
  // raw string and drive off idx_send_stats_due. Previously this used
  // datetime('now') (SQLite's " " separator), which forced a datetime()
  // wrapper in the due query, made the index non-sargable, and regressed the
  // plan to a full sends scan (~9k rows/tick).
  const now = nowISO();
  return db
    .prepare(
      `UPDATE send_stats SET next_fetch_at = ?, updated_at = ?
         WHERE send_id = ? AND next_fetch_at IS NULL AND is_final = 0`,
    )
    .bind(now, now, sendID);
}

export function incrementSendStatsUnsubscribedStmt(
  db: D1Database,
  sendID: string,
): D1PreparedStatement {
  return db
    .prepare(
      `INSERT INTO send_stats (send_id, unsubscribed, updated_at)
         VALUES (?, 1, datetime('now'))
         ON CONFLICT(send_id) DO UPDATE SET
           unsubscribed = unsubscribed + 1,
           updated_at = datetime('now')`,
    )
    .bind(sendID);
}

export type ApplyMailgunStatsParams = {
  sent: number;
  delivered: number;
  opened: number;
  clicked: number;
  failed: number;
  complained: number;
  next_fetch_at: string | null;
  is_final: boolean;
  fetch_error?: string | null;
};

export async function recordStatsFetchError(
  db: D1Database,
  sendID: string,
  nextFetchAt: string | null,
  isFinal: boolean,
  errMsg: string,
): Promise<void> {
  const now = nowISO();
  await db
    .prepare(
      `UPDATE send_stats SET
         last_fetched_at = ?,
         next_fetch_at = ?,
         is_final = ?,
         fetch_error = ?,
         updated_at = ?
       WHERE send_id = ?`,
    )
    .bind(now, isFinal ? null : nextFetchAt, isFinal ? 1 : 0, errMsg, now, sendID)
    .run();
}

export async function applyMailgunStats(
  db: D1Database,
  sendID: string,
  p: ApplyMailgunStatsParams,
): Promise<void> {
  const now = nowISO();
  await db
    .prepare(
      `UPDATE send_stats SET
         sent = ?, delivered = ?, opened = ?, clicked = ?, failed = ?, complained = ?,
         first_fetched_at = COALESCE(first_fetched_at, ?),
         last_fetched_at = ?,
         next_fetch_at = ?,
         is_final = ?,
         fetch_error = ?,
         updated_at = ?
       WHERE send_id = ?`,
    )
    .bind(
      p.sent,
      p.delivered,
      p.opened,
      p.clicked,
      p.failed,
      p.complained,
      now,
      now,
      p.is_final ? null : p.next_fetch_at,
      p.is_final ? 1 : 0,
      p.fetch_error ?? null,
      now,
      sendID,
    )
    .run();
}

export async function listDueSendStats(
  db: D1Database,
  limit: number,
): Promise<DueStatsRow[]> {
  const { results } = await db
    .prepare(
      // Shape matters more than it looks. Two constraints, both learned the
      // hard way after this query burned 27M rows/day (the entire D1 daily
      // read allowance) by full-scanning sends every tick:
      //
      //  1. next_fetch_at is written in ONE format everywhere (JS ISO-8601 via
      //     nowISO), so a raw lexical "<=" is exactly chronological and keeps
      //     idx_send_stats_due sargable. The old datetime()-wrapped compare was
      //     non-sargable. Migration 0015 backfills legacy SQLite-format rows.
      //  2. The due set is resolved in a LIMITed subquery over send_stats
      //     ALONE. SQLite cannot flatten a subquery that has LIMIT, so it must
      //     materialise <=limit rows first and can never choose sends as the
      //     driving table. Do NOT rewrite this as a flat two-table join: D1
      //     does not honour ANALYZE statistics, so a flat join is free to pick
      //     the catastrophic plan and there is no way to talk it out of it.
      `SELECT s.id AS send_id, s.created_at, s.completed_at
         FROM (
           SELECT send_id, next_fetch_at
             FROM send_stats
            WHERE is_final = 0
              AND next_fetch_at IS NOT NULL
              AND next_fetch_at <= ?
            ORDER BY next_fetch_at ASC
            LIMIT ?
         ) due
         JOIN sends s ON s.id = due.send_id
        WHERE s.completed_at IS NOT NULL
        ORDER BY due.next_fetch_at ASC`,
    )
    .bind(nowISO(), limit > 0 ? limit : 25)
    .all<DueStatsRow>();
  return results;
}

File: src/internal/store/stats.go
package store

import (
	"context"
	"database/sql"
	"errors"
	"time"

	"github.com/ranaroussi/minigun/internal/models"
)

func (s *Store) GetSendStats(ctx context.Context, sendID string) (*models.SendStats, error) {
	row := s.DB.QueryRowContext(ctx, `
		SELECT send_id, sent, delivered, opened, clicked, failed, complained, unsubscribed,
		       first_fetched_at, last_fetched_at, next_fetch_at, is_final, fetch_error,
		       created_at, updated_at
		FROM send_stats WHERE send_id = ?`, sendID)
	return scanSendStats(row)
}

func scanSendStats(row *sql.Row) (*models.SendStats, error) {
	var st models.SendStats
	var firstFetched, lastFetched, nextFetch, fetchErr sql.NullString
	var createdAt, updatedAt string
	var isFinalInt int
	if err := row.Scan(
		&st.SendID, &st.Sent, &st.Delivered, &st.Opened, &st.Clicked, &st.Failed, &st.Complained, &st.Unsubscribed,
		&firstFetched, &lastFetched, &nextFetch, &isFinalInt, &fetchErr,
		&createdAt, &updatedAt,
	); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, ErrNotFound
		}
		return nil, err
	}
	st.IsFinal = isFinalInt == 1
	st.FetchError = stringPtr(fetchErr)
	var err error
	if st.FirstFetchedAt, err = parseTimePtr(firstFetched); err != nil {
		return nil, err
	}
	if st.LastFetchedAt, err = parseTimePtr(lastFetched); err != nil {
		return nil, err
	}
	if st.NextFetchAt, err = parseTimePtr(nextFetch); err != nil {
		return nil, err
	}
	if st.CreatedAt, err = parseTime(createdAt); err != nil {
		return nil, err
	}
	if st.UpdatedAt, err = parseTime(updatedAt); err != nil {
		return nil, err
	}
	return &st, nil
}

func (s *Store) initSendStatsRow(ctx context.Context, tx *sql.Tx, sendID string) error {
	_, err := tx.ExecContext(ctx,
		`INSERT OR IGNORE INTO send_stats (send_id) VALUES (?)`, sendID,
	)
	return err
}

type SendStatsUpdate struct {
	Sent       uint64
	Delivered  uint64
	Opened     uint64
	Clicked    uint64
	Failed     uint64
	Complained uint64
	NextFetch  *time.Time
	IsFinal    bool
	FetchError *string
}

func (s *Store) RecordStatsFetchError(ctx context.Context, sendID string, nextFetch *time.Time, isFinal bool, errMsg string) error {
	now := nowISO()
	var nextFetchVal any
	if nextFetch != nil && !isFinal {
		nextFetchVal = nextFetch.UTC().Format(time.RFC3339Nano)
	}
	isFinalInt := 0
	if isFinal {
		isFinalInt = 1
	}
	_, err := s.DB.ExecContext(ctx, `
		UPDATE send_stats SET
		  last_fetched_at = ?,
		  next_fetch_at = ?,
		  is_final = ?,
		  fetch_error = ?,
		  updated_at = ?
		WHERE send_id = ?`,
		now, nextFetchVal, isFinalInt, errMsg, now, sendID,
	)
	return err
}

func (s *Store) ApplyMailgunStats(ctx context.Context, sendID string, u SendStatsUpdate) error {
	now := nowISO()
	var nextFetch any
	if u.NextFetch != nil && !u.IsFinal {
		nextFetch = u.NextFetch.UTC().Format(time.RFC3339Nano)
	}
	isFinalInt := 0
	if u.IsFinal {
		isFinalInt = 1
	}
	_, err := s.DB.ExecContext(ctx, `
		UPDATE send_stats SET
		  sent = ?, delivered = ?, opened = ?, clicked = ?, failed = ?, complained = ?,
		  first_fetched_at = COALESCE(first_fetched_at, ?),
		  last_fetched_at = ?,
		  next_fetch_at = ?,
		  is_final = ?,
		  fetch_error = ?,
		  updated_at = ?
		WHERE send_id = ?`,
		u.Sent, u.Delivered, u.Opened, u.Clicked, u.Failed, u.Complained,
		now, now, nextFetch, isFinalInt, nullString(u.FetchError), now,
		sendID,
	)
	return err
}

func (s *Store) IncrementSendStatsUnsubscribed(ctx context.Context, tx *sql.Tx, sendID string) error {
	if sendID == "" {
		return nil
	}
	exec := func(query string, args ...any) error {
		if tx != nil {
			_, err := tx.ExecContext(ctx, query, args...)
			return err
		}
		_, err := s.DB.ExecContext(ctx, query, args...)
		return err
	}
	return exec(`
		INSERT INTO send_stats (send_id, unsubscribed, updated_at)
		VALUES (?, 1, datetime('now'))
		ON CONFLICT(send_id) DO UPDATE SET
		  unsubscribed = unsubscribed + 1,
		  updated_at = datetime('now')`,
		sendID,
	)
}

func (s *Store) MarkSendCompletedForStats(ctx context.Context, sendID string) error {
	_, err := s.DB.ExecContext(ctx, `
		UPDATE send_stats SET next_fetch_at = datetime('now'), updated_at = datetime('now')
		WHERE send_id = ? AND next_fetch_at IS NULL AND is_final = 0`, sendID,
	)
	return err
}

type DueStatsRow struct {
	SendID      string
	CreatedAt   time.Time
	CompletedAt time.Time
}

func (s *Store) ListDueSendStats(ctx context.Context, limit int) ([]DueStatsRow, error) {
	if limit <= 0 {
		limit = 50
	}
	// next_fetch_at is written in two formats: ApplyMailgunStats stores
	// RFC3339Nano ("2026-06-05T12:09:15.905Z") while MarkSendCompletedForStats
	// stores SQLite's datetime() ("2026-06-05 12:09:15"). A raw string "<="
	// compares the 'T' separator (0x54) against ' ' (0x20), so an RFC3339
	// timestamp always sorts AFTER a same-day SQLite "now" and the row never
	// reads as due until the next calendar day. Normalize both sides through
	// datetime() so the comparison is chronological.
	rows, err := s.DB.QueryContext(ctx, `
		SELECT s.id, s.created_at, s.completed_at
		FROM send_stats ss
		JOIN sends s ON s.id = ss.send_id
		WHERE ss.is_final = 0
		  AND ss.next_fetch_at IS NOT NULL
		  AND datetime(ss.next_fetch_at) <= datetime('now')
		  AND s.completed_at IS NOT NULL
		ORDER BY datetime(ss.next_fetch_at) ASC
		LIMIT ?`, limit,
	)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []DueStatsRow
	for rows.Next() {
		var r DueStatsRow
		var created, completed string
		if err := rows.Scan(&r.SendID, &created, &completed); err != nil {
			return nil, err
		}
		if r.CreatedAt, err = parseTime(created); err != nil {
			return nil, err
		}
		if r.CompletedAt, err = parseTime(completed); err != nil {
			return nil, err
		}
		out = append(out, r)
	}
	return out, rows.Err()
}
